File: C/DataStructures/Linear/Queue/LinkedListBased/queue.c
#include<stdio.h>
#include<stdlib.h>

struct Node{
    int data;
    struct Node* next;
} *start = NULL;

typedef struct Node* node;

int isEmpty(){
    return start == NULL;
}

void insert( int value ){
    node new_node = (node)malloc(sizeof(node));
    new_node -> data = value;
    new_node -> next = NULL;
    if( isEmpty() )
        start = new_node;
    else{
        node ptr = start;
        while( ptr -> next != NULL )
            ptr = ptr -> next;
        ptr -> next = new_node;
    }
}

void deQueue(){
    node ptr = start;
    if( isEmpty() )
        printf("THE QUEUE LIST IS EMPTY.");
    else{
        printf("THE DELETED ELEMENT IS %d", start -> data);
        start = start -> next;
        free(ptr);
    }
}

void display(){
    int i;
    node ptr;
    while( ptr -> next != NULL ){
        ptr = ptr -> next;
        i++;
    }
    printf("%d", i);
}

int search( int value){
    int search_var = 0;
    node ptr = start;
    while( ptr != NULL ){
        if( ptr == value ){
            search_var = 1;
            break;
        }
        else
            ptr = ptr -> next;
    return search_var == 1;
    }
}

int size( int count ){
    node ptr = start;
    while( ptr -> next != NULL ){
        count++;
        ptr = ptr -> next;
    }
    return count;
}
int main(){
    int choice, input;
    while(1){
        printf("\n\n\tQUEUE LINKED LIST.\t\n");
        printf("\n1.ENTER TO INSERT AN ELEMENT TO THE LINKED LIST.");
        printf("\n2.ENTER TO DELETE AN ELEMENT TO THE LINKED LIST.");
        printf("\n3.ENTER TO DISPLAY ALL ELEMNETS IN THE LINKED LIST.");
        printf("\n4.ENTER TO CHECK WHETHER THE LINKED LIST IS EMPTY OR NOT.");
        printf("\n5.ENTER TO SEARCH AN ELEMENT IN THE LINKED LIST.");
        printf("\n6.ENTER TO CHECK THE SIZE OF THE LINKED LIST.");
        printf("\n7.EXIT.");
        printf("\n\nENTER YOUR CHOICE :");
        scanf("%d", &choice);

        switch ( choice ){
            
            case 1 :        printf("ENTER THE ELEMENT TO BE INSERTED :");
                            scanf("%d", &input);
                            insert(input);
            break;

            case 2 :    if( isEmpty() )
                            printf("THE QUEUE IS EMPTY...CAN'T DELETE ANY ELEMENT.");
                        else
                            deQueue();
            break;

            case 3 :    if( isEmpty() )
                            printf("THERE ARE NO ELEMENSTS IN THE LINKED LIST.");
                        else
                            display();
            break;
            
            case 4 :    if( isEmpty() )
                            printf("THE QUEUE IS EMPTY.");
                        else
                            printf("THE QUEUE IS NOT EMPTY.");

            case 5 :        printf("ENTER THE ELEMENT TO BE SEARCHED FOR :");
                            scanf("%d", &input);
                            node ptr;
                        if( ptr -> data == input )
                            printf("THE ELEMRNT IS IN THE LIST.");
                        else
                            printf("THE ELEMENT IS NOT IN THE LIST.");
            break;

            case 6 :    if( isEmpty() )
                            printf("THE SIZE OF THE LIST IS ZERO.");
                        else
                            printf("THE SIZE OF THE LIST IS %d", size(input) );
            break;
            
            case 7 :    exit(1);

            default :   printf("ENTER THE VALID CHOICE.");
        }
    }
    return 0;
}

File: Java/DS/Queues/Queue.java
import java.util.Scanner;

class Operations {
    protected int front, rear, N;
    protected int[] queue;

    Operations( int size ) {
        N = size;
        front = -1;
        rear = -1;
        queue = new int[N];
    }

    public boolean isEmpty() {
        return front == rear;
    }

    public boolean isFull() {
        if( rear == N )
            return true;
        else
            return false;
    }

    public void enQueue( int value ) {
        if( isEmpty() ) 
            front = rear = 0;
        queue[rear] = value;
        rear++;
    }

    public void deQueue() {
        System.out.print("THE DELETED ELEMENT IS " + queue[front]);
        front++;
    }

    public void display() {
        for( int i = front; i < rear; ++i ) 
            System.out.print("->" + queue[i] + "\t");
    }
}

public class Queue {
    public static void main( String[] args ) {
        Scanner scan = new Scanner(System.in);

        System.out.print("ENTER THE MAXIMUN LIMIT OF THE ARRAY : ");
        int N = scan.nextInt();
        int[] queue = new int[N];

        Operations O = new Operations(N);

        while( true ) {
            System.out.print("\n\n\tQUEUE\t\n\n");
            System.out.println("1.ENTER TO INSERT INTO THE QUEUE.");
            System.out.println("2.ENTER TO DELETE FROM THE QUEUE.");
            System.out.println("3.ENTER TO DISPLAY ALL THE ELEMENTS IN THE QUEUE.");
            System.out.println("4.ENTER TO CHECK WHETHER THE QUEUE IS EMPTY OR NOT.");
            System.out.println("5.ENTER TO CHECK WHETHER IS QUEUE IS FULL OR NOT.");
            System.out.print("6.EXIT.");
            System.out.print("\n\nENTER YOUR CHOICE : ");

            switch( scan.nextInt() ) {

                case 1 :    if( O.isFull() )
                                System.out.print("QUEUE OVERFLOW!");

                            else {
                                System.out.print("ENTER THE ELEMENT TO BE INSERTED : ");
                                O.enQueue( scan.nextInt() );
                            }

                break;

                case 2 :    if( O.isEmpty() ) 
                                System.out.print("QUEUE UNDERFLOW!");

                            else
                                O.deQueue();

                break;

                case 3 :    if( O.isEmpty() )
                                System.out.print("QUEUE UNDERFLOW!");

                            else 
                                O.display();

                break;

                case 4 :    if( O.isEmpty() )
                                System.out.print("THE QUEUE IS EMPTY.");

                            else
                                System.out.print("THE QUEUE IS NOT EMPTY.");

                break;

                case 5 :    if( O.isFull() ) 
                                System.out.print("THE QUEUE IS FULL.");

                            else 
                                System.out.print("THE QUEUE IS NOT FULL.");

                break;

                case 6 :    System.exit(1);

                default :   System.out.print("ENTER A VALID CHOICE.");
            }
        }
    }
}